src/Data/Group/Order.hs

```diff
@@ -0,0 +1,190 @@
+--   
+-- /Note:/ For any valid instances of 'GroupOrder',
+-- @order x == Finite 1@ holds if and only if @x == mempty@.
+-- 
+-- === __Examples__:
+-- 
+-- >>> order (3 :: Sum Word8)
+-- Finite 256
+-- >>> order (16 :: Sum Word8)
+-- Finite 16
+-- >>> order (0 :: Sum Integer)
+-- Finite 1
+-- >>> order (1 :: Sum Integer)
+-- Infinite
+class (Eq g, Group g) => GroupOrder g where
+    -- | The order of an element of a group.
+    --   
+    -- @order x@ must be @Finite k@ if the order of @x@ is
+    -- finite @k@, and must be @Infinite@ otherwise.
+    --   
+    -- For a type which is also 'FiniteGroup',
+    -- @'Finite' . 'finiteOrder'@ is a valid implementation of 'order',
+    -- if not efficient.
+    order :: g -> Order
+
+instance GroupOrder () where
+    order _ = Finite 1
+
+instance GroupOrder (Proxy a) where
+    order _ = Finite 1
+
+instance GroupOrder (Sum Integer) where
+    order 0 = Finite 1
+    order _ = Infinite
+
+instance GroupOrder (Sum Rational) where
+    order 0 = Finite 1
+    order _ = Infinite
+
+instance GroupOrder (Sum Int) where order = orderForBits
+instance GroupOrder (Sum Int8) where order = orderForBits
+instance GroupOrder (Sum Int16) where order = orderForBits
+instance GroupOrder (Sum Int32) where order = orderForBits
+instance GroupOrder (Sum Int64) where order = orderForBits
+instance GroupOrder (Sum Word) where order = orderForBits
+instance GroupOrder (Sum Word8) where order = orderForBits
+instance GroupOrder (Sum Word16) where order = orderForBits
+instance GroupOrder (Sum Word32) where order = orderForBits
+instance GroupOrder (Sum Word64) where order = orderForBits
+
+-- | Given a number @x :: a@ represented by fixed-width binary integers,
+--   return the minimum positive integer @2^n@ such that
+--   @(fromInteger (2^n) * x :: a) == 0@.
+zeroFactor :: FiniteBits a => a -> Natural
+zeroFactor a = bit (finiteBitSize a - countTrailingZeros a)
+
+-- | An efficient implementation of 'order' for additive group of
+--   fixed-width integers, like 'Int' or 'Word8'.
+orderForBits :: (Integral a, FiniteBits a) => Sum a -> Order
+orderForBits (Sum a) = Finite (zeroFactor a)
+
+instance GroupOrder (Product Rational) where
+    order 1 = Finite 1
+    order _ = Infinite
+
+instance (GroupOrder a, GroupOrder b) => GroupOrder (a,b) where
+    order (a,b) = order a `lcmOrder` order b
+
+instance (GroupOrder a, GroupOrder b, GroupOrder c) => GroupOrder (a,b,c) where
+    order (a,b,c) = order ((a,b),c)
+
+instance (GroupOrder a, GroupOrder b, GroupOrder c, GroupOrder d)
+        => GroupOrder (a,b,c,d) where
+    order (a,b,c,d) = order ((a,b),(c,d))
+instance (GroupOrder a, GroupOrder b, GroupOrder c, GroupOrder d, GroupOrder e)
+        => GroupOrder (a,b,c,d,e) where
+    order (a,b,c,d,e) = order ((a,b,c),(d,e))
+
+{- Safe Haskell doesn't allow GND, at least for now.
+{-# language
+  GeneralizedNewtypeDeriving,
+  StandaloneDeriving,
+  DerivingStrategies
+#-}
```

Comment:
I'm not tied to safe, if you want to mark this as `-XTrustworthy` and make things easier.

Reply:
Eh, we're good. We should dedicate a PR to that if we need it.